### PR TITLE
gsoc26: Implementing canonical pipeline for deterministic json

### LINKED ISSUE #1067: (GSoC'26) Week 1: Improve and fix canonical.js and importCanonical.js for all circuitElementList types.
Description:

Implement canonical.ts : This is the file that will convert the circuit to the canonical JSON.
Implement the Weisfeiler-Lehman algorithm in canonical.ts for determinism.
Determinism should be achieved for all circuit types and all of their properties.

### PR DESCRIPTION
**Pull request by Harkeerat24**
Fixes #1067

 

#### Describe the changes you have made in this PR -
- Introduces `canonical.ts` — a deterministic serialisation pipeline for saving circuits.

## The core problem this solves
- Two circuits that are structurally identical but were built in different orders, or have components placed at different coordinates, would previously produce different save files with no way to compare them programmatically. This PR lays the foundation for a stable canonical representation.

### Screenshots of the UI changes (If any) -

No changes in UI
---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [x] No, I wrote all the code myself
- [ ] Yes, I used AI assistance (continue below)

**Explain your implementation approach:**

- **`UnionFind`** — path-compressed, union-by-rank disjoint-set structure used to group all wire nodes that are electrically connected into named nets.

- **Net discovery pipeline** — `indexNodes` assigns a numeric index to every `WireNode` in `scope.allNodes`; `discoverNets` then unions connected pairs so every wire segment belonging to the same net shares one representative root index.

- **Component draft builder** (`buildComponentDrafts`) — walks `circuitElementList` to collect every placed component, calls `customSave()` on each to retrieve port definitions, resolves each port's node to its net root via UnionFind, and captures
  propagation delay, constructor parameters, and stateful default values into a flat `ComponentDraft` record.

- **1-WL graph fingerprinting** — `wlFingerprint` runs the Weisfeiler-Leman colour refinement algorithm. Components sharing the same type,  bitWidth, and port-arity start with the same colour; repeated neighbourhood aggregation then distinguishes them based on structural context. This produces a stable relative ordering that is independent of insertion order or canvas position.

- **`canonicalSort`** — uses the WL colours as a tiebreaker on top of a `(type, bitWidth, portKey)` primary key, then sorts components in-place with `naturalCompare` (locale-free, digit-aware) so IDs assigned next are deterministic.

- **Net and component ID assignment** (`assignComponentIds`, `assignNetIds`, `buildCanonicalNets`) — assigns `Type_N` IDs to components in sorted order, then traverses ports in natural order to mint `net_N` IDs; single-endpoint nets (dangling wires) are discarded.

- **Wire junction layout** (`buildWireJunctions`) — records intermediate junction nodes (type-2 nodes) per net with their local coordinates and adjacency edges, so the canonical file can reconstruct wire routing without embedding raw node indices.

- **`canonicaliseScope`** — orchestrates the full pipeline, strips labels and nulls direction/layout params before hashing to avoid cosmetic differences affecting the structural hash, then computes a SHA-256 digest via `crypto.subtle`.

- You can use `await (await import('/simulatorvue/v1/v1/src/simulator/src/data/canonical.ts')).canonicaliseScope(globalScope)` in the console to print the JSON as shown in the figure below:

image

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [ ] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

## Summary by CodeRabbit

* **New Features**
  * Introduced a canonicalization pipeline that converts simulator circuit scopes into a deterministic representation, including stable component ordering, net identification, junction reconstruction, and generated layout metadata.
  * Added a determinism hash for canonical netlists and updated browser availability via a global helper.

### CODE REVIEW

File: v1/src/simulator/src/data/canonical.ts [hunk 1]
```diff
@@ -0,0 +1,891 @@
+import { circuitElementList } from "../metadata";
```
[the remark this answers is not in the document]
Comment (Harkeerat24) [reply v1/src/simulator/src/data/canonical.ts:1]: This is intentionally placed in the v1 folder to safely test the experimental pipeline separately without affecting the stable v0

**Comment by aryanndwi123** [comment]
@JoshVarga @Arnabdaz @tachyons I've completed my review. Could you please take a look as well so we can move forward with merging it?

**Comment by aryanndwi123** [comment]
LGTM 🚀

**Review by JoshVarga — APPROVED** [review]

**Comment by Harkeerat24** [comment]
#1095 fixes this, closing this PR
